### PR TITLE
Fixing bugs in the computation of implicit arguments for `Fixpoint` with a let binder

### LINKED ISSUE #3282: VM confused by let bindings in fixpoints
Note: the issue was created automatically with bugzilla2github tool

Original bug ID: BZ#3282
From: @maximedenes
Reported version: 8.5
CC:   @herbelin, @JasonGross

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** bug fix

Fixes / closes #3282 (it is about extra bugs related to #3282).

- [X] Added / updated test-suite
- [X] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

File: doc/changelog/02-specification-language/11132-master+fix-implicit-let-fixpoint-bug3282.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+- Bugs fixed in computing implicit arguments for fix/cofix in the presence of let-in, see #3282 (`#11132 <https://github.com/coq/coq/pull/11132>`_, by Hugo Herbelin).
```
Comment (Zimmi48) [inline doc/changelog/02-specification-language/11132-master+fix-implicit-let-fixpoint-bug3282.rst]: Does this really justify a changelog entry? It's just an anomaly that disappears IIUC.

Comment (herbelin) [reply doc/changelog/02-specification-language/11132-master+fix-implicit-let-fixpoint-bug3282.rst]: It's more than an anomaly, it is also some (co)fixpoints with implicit arguments which were wrongly rejected. But I have no good intuition yet on when a changelog is expected and I can remove it if you want.

Comment (Zimmi48) [reply doc/changelog/02-specification-language/11132-master+fix-implicit-let-fixpoint-bug3282.rst]: I would rather say that your changelog entries are (similar to your PR title and commit messages) often unclear. In this case, I would have preferred something like "Fixed a bug preventing to define some valid (co)fixpoints with implicit arguments."

Comment (herbelin) [reply doc/changelog/02-specification-language/11132-master+fix-implicit-let-fixpoint-bug3282.rst]: Ah, sorry. Thanks for telling. Your formulation is nice and I took it as a basis for a reformulation.

For the change logs, will there be a manual pass over them before release to harmonize them and/or removing the one which appear at the end not relevant enough?

Comment (Zimmi48) [reply doc/changelog/02-specification-language/11132-master+fix-implicit-let-fixpoint-bug3282.rst]: Yes, this is the role of the RM but something that I help with given the current state that is still somewhat experimental.

**Review by ejgallego — APPROVED** [review]
Fix looks correct thanks; and indeed IMO this deserves a changelog entry, I would even illustrate an example of after / before which I think would be the most illustrative for the users.

@herbelin you should rebase to make the linter happy [if you don't mind, sorry for the hassle]

File: doc/changelog/02-specification-language/11132-master+fix-implicit-let-fixpoint-bug3282.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+- Fixed bugs sometimes preventing to define valid (co)fixpoints with implicit arguments in the presence of local definitions, see #3282 (`#11132 <https://github.com/coq/coq/pull/11132>`_, by Hugo Herbelin).
```
Comment (Zimmi48) [inline doc/changelog/02-specification-language/11132-master+fix-implicit-let-fixpoint-bug3282.rst]: The suggestion is mostly to add a link:

```suggestion
- Fixed bugs sometimes preventing to define valid (co)fixpoints with implicit arguments
  in the presence of local definitions, see `#3282 <https://github.com/coq/coq/issues/3282>`_
  (`#11132 <https://github.com/coq/coq/pull/11132>`_, by Hugo Herbelin).
```

I agree with Emilio. Including an example in the changelog would be great. Feel free to use the `.. example::` and the `.. coqtop::` directives.

**Comment by ejgallego** [comment]
Indeed @Zimmi48 suggestion seems like worth adding; will wait a bit more to let @herbelin take it into account if he considers it worthwhile.

**Comment by herbelin** [comment]
@Zimmi48, @ejgallego: I made an attempt as using `example::` and `coqtop::`. I hope I did it correctly, so please confirm. 

If the RM finds this too verbose or even too minor, no problem if the message is shortcut or discarded.

**Comment by ejgallego** [comment]
> @Zimmi48, @ejgallego: I made an attempt as using `example::` and `coqtop::`. I hope I did it correctly, so please confirm.
> 
> If the RM finds this too verbose or even too minor, no problem if the message is shortcut or discarded.

Seems to work fine, thanks @herbelin ; I like the changelog entry.

**Review by Zimmi48 — APPROVED** [review]
Great!

**Comment by vbgl** [comment]
Is this really for 8.10.2? In that case I would need some help for backporting. Thanks.

**Comment by Zimmi48** [comment]
Likely not critical enough to bother backporting to 8.10 if the backporting is not straightforward.

**Comment by herbelin** [comment]
Not very critical indeed. Likely to remain unobserved in practice.

**Pull request merged**